src/org/labkey/test/tests/SampleTypeLinkToStudyTest.java

```diff
@@ -0,0 +1,194 @@
+        dgen.insertRows(createDefaultConnection(), dgen.getRows());
+
+        lookupInfo = new FieldDefinition.LookupInfo(getProjectName(), "exp.materials", SAMPLE_TYPE2);
+        dgen = new TestDataGenerator(lookupInfo)
+                .withColumns(List.of(
+                        new FieldDefinition("name", FieldDefinition.ColumnType.String),
+                        new FieldDefinition("int", FieldDefinition.ColumnType.Integer)));
+        dgen.createDomain(createDefaultConnection(), SAMPLE_TYPE_DOMAIN_KIND);
+        dgen.addCustomRow(Map.of("name", "first"));
+        dgen.addCustomRow(Map.of("name", "second"));
+        dgen.addCustomRow(Map.of("name", "third"));
+        dgen.addCustomRow(Map.of("name", "fourth"));
+        dgen.insertRows(createDefaultConnection(), dgen.getRows());
+
+    }
+
+    @Test
+    public void testLinkToStudy()
+    {
+        goToProjectHome(SAMPLE_TYPE_PROJECT);
+
+        log("Linking sample types one row to study");
+        linkToStudy(SAMPLE_TYPE1, 2, Arrays.asList("P1", "P2"), Arrays.asList("1", "2"));
+
+        log("Verifying the linked sample type in study");
+        goToProjectHome(LINKED_STUDY_PROJECT);
+        clickAndWait(Locator.linkWithText(SAMPLE_TYPE1));
+        DataRegionTable table = new DataRegionTable("Dataset", getDriver());
+
+        checker().verifyEquals("Incorrect number of rows linked", 2, table.getDataRowCount());
+        checker().verifyEquals("Incorrect Participant ID's", Arrays.asList("P1", "P2"), table.getColumnDataAsText("ParticipantId"));
+
+        log("Verifying log entries");
+        goToProjectHome(SAMPLE_TYPE_PROJECT);
+        clickAndWait(Locator.linkWithText(SAMPLE_TYPE1));
+        String expectedComment = "2 row(s) were linked to a study from the sample type: " + SAMPLE_TYPE1;
+        verifyLinkToHistory(expectedComment);
+        verifyAuditLogEvents(expectedComment);
+    }
+
+    @Test
+    public void testDatasetRecall()
+    {
+        goToProjectHome(SAMPLE_TYPE_PROJECT);
+        linkToStudy(SAMPLE_TYPE2, 1, Arrays.asList("P3"), Arrays.asList("3"));
+
+        goToProjectHome(LINKED_STUDY_PROJECT);
+        clickAndWait(Locator.linkWithText(SAMPLE_TYPE2));
+        DataRegionTable table = new DataRegionTable("Dataset", getDriver());
+        table.checkCheckbox(0);
+        table.clickHeaderButton("Recall");
+        acceptAlert();
+        checker().verifyEquals("Dataset row not recalled", 0, table.getDataRowCount());
+
+        log("Verifying log entries");
+        goToProjectHome(SAMPLE_TYPE_PROJECT);
+        clickAndWait(Locator.linkWithText(SAMPLE_TYPE2));
+        String expectedComment = "1 row(s) were recalled from a study to the sample type: " + SAMPLE_TYPE2;
+        verifyLinkToHistory(expectedComment);
+        verifyAuditLogEvents(expectedComment);
+    }
+
+    private void linkToStudy(String sampleName, int numOfRowsToBeLinked, List<String> participantID, List<String> visitId)
+    {
```

Comment:
Not a big deal but it looks like numOfRowsToBeLinked is equal to participantID.size() or VisitId.size(), so you don't technically need that parameter.

You could also (if you wanted) merge participantID and VisitId into a Map<String, String> where the participantID is the key. It doesn't do a lot, just an alternative way to implement the function.

I would suggest more consistent variable naming: participant**ID** vs visit**Id**.

Reply:
Changed the method signature itself. Thank you